### PR TITLE
Enhanced load verify locations to support flags

### PR DESCRIPTION
**Pull request by dgarske**
* Fixed `wolfSSL_CTX_load_verify_locations` to continue loading if there is an error (ZD 4265).
* Added new `wolfSSL_CTX_load_verify_locations_ex` that supports flags `WOLFSSL_LOAD_FLAG_IGNORE_ERR`, `WOLFSSL_LOAD_FLAG_DATE_ERR_OKAY` and `WOLFSSL_LOAD_FLAG_PEM_CA_ONLY`.
* Fix for `PemToDer` to handle PEM which may include a null terminator in length at end of file length causing wrong error code to be returned. Added test case for this. (ZD 4278) 
* Added macro to override default flags for `wolfSSL_CTX_load_verify_locations` using `WOLFSSL_LOAD_VERIFY_DEFAULT_FLAGS`.
* Added tests for loading CA PEM's from directory using `wolfSSL_CTX_load_verify_locations` and `wolfSSL_CTX_load_verify_locations_ex` with flags.
* Added tests for `wolfSSL_CertManagerLoadCABuffer`.
* Updated the expired test certs and added them to `./certs/test/gen-testcerts.sh` script.

### CODE REVIEW

**Comment by toddouska** [comment]
Please justify completely changing the return semantics to succeed when all loads fail.  Or some loads fail.

**Comment by toddouska** [comment]
I think continuing is OK on a failed cert, in fact, I remember we worked on that a few years ago (maybe it got lost in a cleanup or was file based vs dir based).

I'm still inclined to return a failure code if everything doesn't load correctly.  Imagine you have a PKI with 2 root certs.  You call the API to load them but one fails and you get a success return code, later some connections keep failing with unknown root and the user can't understand it.

What do you think about that case?

File: src/ssl.c [hunk 1]
```diff
@@ -5769,28 +5769,60 @@ int wolfSSL_CTX_load_verify_locations(WOLFSSL_CTX* ctx, const char* file,
...
         wc_ReadDirClose(readCtx);
 
         /* pass directory read failure to response code */
-        if (ret == WOLFSSL_SUCCESS && fileRet != -1) {
+        if (fileRet != WC_READDIR_NOFILE) {
             ret = fileRet;
         }
+        else if (file == NULL && successCount == 0) {
+            /* report failure if no files were loaded */
+            ret = WOLFSSL_FAILURE;
```
Comment (toddouska) [inline src/ssl.c]: This still appears to succeed if one of two certs are ok, that is, one fails.

File: src/ssl.c [hunk 2]
```diff
@@ -5767,30 +5768,70 @@ int wolfSSL_CTX_load_verify_locations(WOLFSSL_CTX* ctx, const char* file,
...
             ret = fileRet;
         }
+        else if (file == NULL && (successCount == 0 || failCount > 0)) {
+            /* report failure if no files were loaded or there were failures */
+            /* pass on existing error code if exists */
+            if (ret == WOLFSSL_SUCCESS)
+                ret = WOLFSSL_FAILURE;
+        }
+        else {
+            ret = WOLFSSL_SUCCESS;
```
Comment (toddouska) [inline src/ssl.c]: This doesn't look right.  If file points to a file and everything in path fails we return success?

Comment (dgarske) [reply src/ssl.c]: @toddouska : This logic has been improved... Thanks

**Pull request merged**